### ISSUE TITLE
tmsu merge A B : 'could not delete tag A: too many SQL variables'

### ISSUE THREAD

**Issue by 0ion9** [oniony/TMSU#32]
I'm not sure what is going on here yet, but this just occurred with tags having
738 and 758 taggings respectively, when merging the smaller into the larger.

The result is that all of the files tagged A get tagged B, and all of the files tagged A are no longer tagged A, but the tag A is not deleted. You can however delete it manually.

I've created a test script which successfully reproduces the problem:

```
#!/bin/sh
mkdir -p /tmp/tmsumerge
cd /tmp/tmsumerge
tmsu -v init
for V in $(seq 1 1000); do echo $V > $V; tmsu -v tag --tags=taga $V ; done
for V in $(seq 1001 1750); do echo $V > $V; tmsu -v tag --tags=tagb $V ; done
echo
echo
echo
# will fail
tmsu -v merge tagb taga
echo
echo taga count
tmsu -v files --count taga
echo tagb count
tmsu -v files --count tagb
```

TMSU version: git 62adf28 (performance improvements to VFS)

**Comment by oniony** [comment]
Thanks for reporting. I'll check it out this evening. I imagine I've forgotten to commit a transaction.

**Comment by oniony** [comment]
OK, tracked this down: issue was with the creation of a truly massive SQL statement when cleaning untagged files up after removal of the old tag.

TMSU always deletes files from the database when they become untagged. Orginally I had some code in place that would check every file whenever any untagging operation was performed but I made it more 'clever' a while back so that it only considered files that could possibly have been left untagged based upon what had just happened. However this 'clever' code was building a massive SQL statement from the set of potentially affected files.

I've changed it to, instead, iterate the set of files and delete each one individually if it is no longer tagged. I have made the same change to the similar code in where unused values are checked too.

However, it does still beg the question: why is it even checking that the files need to be deleted during a merge operation? The nature of the merge means that the same set of files will be tagged at the start and end so the untagged file check can be safely avoided.
